src/main/java/org/openstreetmap/josm/plugins/pt_assistant/actions/PTMendRelationAction.java

```diff
@@ -0,0 +1,2348 @@
+            allowedCharacters.add('R');
+        }
+
+        for (int i = 0; i < 5 && i < wayIndices.size(); i++) {
+            wayColoring.put(members.get(wayIndices.get(i)).getWay(), alphabet);
+        }
+
+        if (notice.equals("vehicle travels against oneway restriction")) {
+            if (numeric) {
+                allowedCharacters.add('4');
+            } else {
+                allowedCharacters.add('C');
+            }
+        }
+
+        // remove any existing temporary layer
+        removeTemporaryLayers();
+
+        if (abort)
+            return;
+
+        // zoom to problem:
+        final Collection<OsmPrimitive> waysToZoom = new ArrayList<>();
+
+        for (Integer i : wayIndices) {
+            waysToZoom.add(members.get(i).getWay());
+        }
+
+        AutoScaleAction.zoomTo(waysToZoom);
+
+        // display the fix variants:
+        temporaryLayer = new MendRelationRemoveLayer();
+        MainApplication.getMap().mapView.addTemporaryLayer(temporaryLayer);
+
+        // // add the key listener:
+        MainApplication.getMap().mapView.requestFocus();
+        MainApplication.getMap().mapView.addKeyListener(new KeyListener() {
+
+            @Override
+            public void keyTyped(KeyEvent e) {
+                // TODO Auto-generated method stub
+            }
+
+            @Override
+            public void keyPressed(KeyEvent e) {
+                downloadCounter = 0;
+                if (abort) {
+                    MainApplication.getMap().mapView.removeKeyListener(this);
+                    MainApplication.getMap().mapView.removeTemporaryLayer(temporaryLayer);
+                    return;
+                }
+                Character typedKey = e.getKeyChar();
+                Character typedKeyUpperCase = typedKey.toString().toUpperCase().toCharArray()[0];
+                if (allowedCharacters.contains(typedKeyUpperCase)) {
+                    nextIndex = true;
+                    MainApplication.getMap().mapView.removeKeyListener(this);
+                    MainApplication.getMap().mapView.removeTemporaryLayer(temporaryLayer);
+                    Logging.debug(String.valueOf(typedKeyUpperCase));
+                    if (typedKeyUpperCase == 'R' || typedKeyUpperCase == '3') {
+                        wayIndices.add(0, currentIndex);
+                    }
+                    RemoveWayAfterSelection(wayIndices, typedKeyUpperCase);
+                }
+                if (e.getKeyCode() == KeyEvent.VK_ESCAPE) {
+                    MainApplication.getMap().mapView.removeKeyListener(this);
+                    Logging.debug("ESC");
+                    nextIndex = false;
+                    setEnable = true;
+                    halt = true;
+                    setEnabled(true);
+                    MainApplication.getMap().mapView.removeTemporaryLayer(temporaryLayer);
+                }
+            }
+
+            @Override
+            public void keyReleased(KeyEvent e) {
```

Comment:
If you override `KeyAdapter` instead of `KeyListener`, you don't need to implement all three methods, but you can just leave out the ones that you don't need.

This is an example of hook methods described in https://en.wikipedia.org/wiki/Template_method_pattern